Lib/test/test_sqlite3/test_cli.py

```diff
@@ -200,5 +205,75 @@ def test_color(self):
 
+
+@requires_subprocess()
+class CompletionTest(unittest.TestCase):
+    PS1 = "sqlite> "
+
+    @classmethod
+    def setUpClass(cls):
+        readline = import_module("readline")
+        if readline.backend == "editline":
+            raise unittest.SkipTest("libedit readline is not supported")
+
+    def write_input(self, input, env=None):
+        script = textwrap.dedent("""
+            import readline
+            readline.parse_and_bind("set colored-completion-prefix off")
+            from sqlite3.__main__ import main; main()
+        """)
+        return run_pty(script, input, env)
+
+    def test_keyword_completion(self):
+        # List candidates starting with 'S', there should be multiple matches.
+        input = b"S\t\tEL\t 1;\n.quit\n"
+        output = self.write_input(input)
+        self.assertIn(b"SELECT", output)
+        self.assertIn(b"SET", output)
+        self.assertIn(b"SAVEPOINT", output)
+        self.assertIn(b"(1,)", output)
+
+        # Keywords are completed in upper case for even lower case user input
+        input = b"sel\t\t 1;\n.quit\n"
+        output = self.write_input(input)
+        self.assertIn(b"SELECT", output)
+        self.assertIn(b"(1,)", output)
+
+    def test_nothing_to_complete(self):
+        input = b"xyzzy\t\t\b\b\b\b\b.quit\n"
+        # set NO_COLOR to disable coloring for self.PS1
+        output = self.write_input(input, env={"NO_COLOR": "1"})
+        output_lines = output.decode().splitlines()
+        line_num = next((i for i, line in enumerate(output_lines, 1)
+                         if line.startswith(f"{self.PS1}xyzzy")), -1)
+        self.assertNotEqual(line_num, -1)
+        # completions occupy lines, assert no extra lines when there is nothing
+        # to complete
+        self.assertEqual(line_num, len(output_lines))
+
+    def test_completion_order(self):
+        script = textwrap.dedent("""
+            import readline
+            readline.parse_and_bind("set colored-completion-prefix off")
+            # hide control sequences surrounding each candidate
+            readline.parse_and_bind("set colored-stats off")
+            # hide "Display all xxx possibilities? (y or n)"
+            readline.parse_and_bind("set completion-query-items 0")
+            # hide "--More--"
+            readline.parse_and_bind("set page-completions off")
+            # show candidates one per line
+            readline.parse_and_bind("set completion-display-width 0")
+            from sqlite3.__main__ import main; main()
+        """)
+        input = b"\t\t.quit\n"
+        output = run_pty(script, input, env={"NO_COLOR": "1"})
+        output_lines = output.decode().splitlines()
+        slices = tuple(i for i, line in enumerate(output_lines) if line.startswith(self.PS1))
+        self.assertEqual(len(slices), 2)
+        start, end = slices
+        candidates = [c.strip() for c in output_lines[start+1 : end]]
```

Comment:
```suggestion
        indices = [i for i, line in enumerate(output_lines)
                   if line.startswith(self.PS1)]
        self.assertEqual(len(indices), 2)
        start, end = indices[0] + 1, indices[1]
        candidates = [c.strip() for c in output_lines[start:end]]
```

I personally always find slice notation with an arithmetic operation inside unreadable due to spaces around `:` due to PEP-8. So I always try to have `i:j` instead of `i+X : j`. It's really a personal matter though but it's hard on me to read. I would have used `list(map(str.strip, output_lines[start:end]))` as well, but I know that pure-functional code is less readable to others.

Reply:
I would have used `map()` as well if I thought of `str.strip` instead of the long `lambda c:c.strip()`. I think `map()` doesn't reduce too much readability but it make the code a bit cleaner.

Found that the `candidates` should not be sorted in `self.assertEqual()`, otherwise the test doesn't achieve its purpose.

Reply:
IMO, this is a misuse of PEP-8. Readability counts. Always. I strongly prefer @tanloong's original code, and I'm going to revert to that before landing.